File: src/2023/day-02/typescript/part-1.ts
export function Part1(input: string) {
  return input
    .trim()
    .split("\n")
    .map((line) => {
      const [label, game] = line.split(": ");
      return game.split("; ").map((set) => {
        return set.split(", ").map((score) => {
          const [v, c] = score.split(" ");
          if (c === "red") return +v <= 12
          else if (c === "green") return +v <= 13
          else if (c === "blue") return +v <= 14
          return false
        })
      }).every((x) => x.every((y) => y)) ? +label.split(" ")[1] : 0
    }).reduce((a, b) => a + b);
}

File: src/2023/day-02/rust/src/part_1.rs
use itertools::Itertools;

pub fn part_1(input: &str) -> i32 {
  input
    .trim()
    .split('\n')
    .map(|line| {
      let (label, games) = line.split(": ").collect_tuple().unwrap();
      games
        .split("; ")
        .map(|set| {
          set
            .split(", ")
            .map(|score| {
              let (v, color) = score.split(' ').collect_tuple().unwrap();
              let v: i32 = v.parse().unwrap();

              match color {
                "red" => v <= 12,
                "green" => v <= 13,
                "blue" => v <= 14,
                _ => panic!("it could be worse"),
              }
            })
            .all(|x| x)
        })
        .all(|x| x)
        .then(|| label.split(' ').nth(1).unwrap().parse::<i32>().unwrap_or(0))
        .unwrap_or(0)
    })
    .sum()
}
